chore: Add new SDK overhead page for Android (#12031)
getsentry/sentry-docs: docs/platforms/android/overhead/index.mdx

---
title: SDK Overhead
description: "Learn about Sentry's Android SDK overhead."
sidebar_order: 7500
---

Adding new features or dependencies to your app incurs additional costs on resources - CPU time, memory usage, network bandwidth, among other things. Sentry SDKs are no different. This document adds transparency to the possible overhead that using this SDK can add, and help you find the feature set and configurations that work best for you.

## General Approach

> The SDK is designed to have minimal to no impact on the user experience. To achieve this, we utilize an array of tools to measure and optimize the performance of our implementations continuously.

Reviewer: ```suggestion
Our SDK is designed to have minimal to no impact on user experience. To achieve this, we utilize an array of tools to continuously measure and optimize the performance of our implementations.
```

We also employ various techniques to ensure we don't add strain on the system's resources along the hot path - on Mobile, this very often means that we offload processing steps, I/O, and other things to a background thread, or we postpone processing to a later time if possible.

If you find (e.g. via local Profiling, or using Sentry to improve the performance of your app) that the SDK does not operate within the guidelines mentioned below, please [open an issue](https://github.com/getsentry/sentry-java/issues/new/choose) on our SDK repo and make sure you provide as much context as you can.

## Error Monitoring

During regular operation, error monitoring incurs little to no overhead. Once an error or crash occurs, the user experience is compromised, and any crash handling routines operate under time constraints imposed by the system. This means that these implementations are highly optimized to perform the required work as quickly as possible.

> The SDK also provides methods to manually capture events: <PlatformLink to="/usage/#capturing-errors">`captureError`</PlatformLink> and <PlatformLink to="/usage/#capturing-messages">`captureMessage`</PlatformLink>. These methods perform some complex operations, such as capturing stack trace information, and while they are highly optimized as well, calling them in tight loops should be avoided.

Reviewer: ```suggestion
Our SDK also provides methods to manually capture events: <PlatformLink to="/usage/#capturing-errors">`captureError`</PlatformLink> and <PlatformLink to="/usage/#capturing-messages">`captureMessage`</PlatformLink>. These methods perform some complex operations, such as capturing stack trace information, and while they are highly optimized as well, calling them in tight loops should be avoided.
```

## Screenshots and View Hierarchy

If you activate these features, the SDK will capture <PlatformLink to="/enriching-events/screenshots">Screenshots</PlatformLink> and <PlatformLink to="/enriching-events/viewhierarchy">View Hierarchy</PlatformLink> of the app's UI at the time of the error crash. This incurs a small overhead that is unnoticeable during normal operation.

If your app raises many errors in a tight loop, it can become too much to process quickly enough, and UI jank can be the result, so make sure you handle such cases appropriately. The SDK provides callbacks to fine-tune when to capture <PlatformLink to="/enriching-events/screenshots#customize-screenshot-capturing">Screenshots</PlatformLink> and <PlatformLink to="/enriching-events/viewhierarchy#customize-view-hierarchy-capturing">View Hierarchy</PlatformLink>, which can also be used to reduce performance impact in scenarios where you don't need screenshots or view hierarchies.

## Breadcrumbs

> <PlatformLink to="/enriching-events/breadcrumbs">Breadcrumbs</PlatformLink> are collected through automated integrations or by manually adding them. To have them readily available for every event generated by the SDK, they are continuously persisted, and managed in a performant buffer. This is expected to have no impact on user experience.

Reviewer: ```suggestion
<PlatformLink to="/enriching-events/breadcrumbs">Breadcrumbs</PlatformLink> are collected through automated integrations or by manually adding them. To have them readily available for every event generated by the SDK, they are continuously persisted, and managed in a performant buffer. This shouldn't impact user experience.
```

> Capturing excessive numbers of breadcrumbs (e.g. by creating breadcrumbs for all log messages) can cause significant performance overhead. To mitigate this, review and adapt your app's usage of breadcrumbs. For example, increase the min-level of log messages for which to create breadcrumbs from `warn` to `error`.

Reviewer: ```suggestion
Capturing excessive numbers of breadcrumbs (for example, creating breadcrumbs for all log messages) can cause significant performance overhead. To mitigate this, review and adapt your app's usage of breadcrumbs. For example, increase the min-level of log messages that create breadcrumbs from `warn` to `error`.
```
